rust/src/delta_datafusion.rs

```diff
@@ -603,11 +606,84 @@ fn left_larger_than_right(left: ScalarValue, right: ScalarValue) -> Option<bool>
     }
 }
 
+/// Responsible for checking batches of data conform to table's invariants.
+pub struct DeltaDataChecker {
+    invariants: Vec<Invariant>,
+    ctx: SessionContext,
+    rt: tokio::runtime::Runtime,
+}
+
+impl DeltaDataChecker {
+    /// Create a new DeltaDataChecker
+    pub fn new(invariants: Vec<Invariant>) -> Self {
+        Self {
+            invariants,
+            ctx: SessionContext::new(),
+            rt: tokio::runtime::Runtime::new().unwrap(),
+        }
+    }
+
+    /// Check that a record batch conforms to table's invariants.
+    ///
+    /// If it does not, it will return [DeltaTableError::InvalidData] with a list
+    /// of values that violated each invariant.
+    pub fn check_batch(&self, record_batch: &RecordBatch) -> Result<(), DeltaTableError> {
+        self.enforce_invariants(record_batch)
+        // TODO: for support for Protocol V3, check constraints
+    }
+
+    fn enforce_invariants(&self, record_batch: &RecordBatch) -> Result<(), DeltaTableError> {
+        // Invariants are deprecated, so let's not pay the overhead for any of this
+        // if we can avoid it.
+        if self.invariants.is_empty() {
+            return Ok(());
+        }
+
+        let table = MemTable::try_new(record_batch.schema(), vec![vec![record_batch.clone()]])?;
+        self.ctx.register_table("data", Arc::new(table))?;
+
+        let mut violations: Vec<String> = Vec::new();
+
+        for invariant in self.invariants.iter() {
```

Comment:
Did you test whether it would be faster to concatenate all invariants into a single filter clause using `OR`?

Reply:
Oh sorry I missed this. I did it separately so that we can provide the user the exact value that violated an invariant. But if it become a performance problem we can re-evaluate. Probably will care more for Constraits than invariants.